CRM/Twingle/Profile.php

```diff
@@ -193,33 +231,76 @@ public function saveProfile() {
-      // existing profile -> just update the config
-      CRM_Core_DAO::executeQuery(
-        "UPDATE civicrm_twingle_profile SET config = %2 WHERE name = %1",
-        [
-          1 => [$this->name, 'String'],
-          2 => [json_encode($this->data), 'String']
-        ]);
-    } else {
-      // new profile -> add new entry to the DB
-      CRM_Core_DAO::executeQuery(
-        "INSERT IGNORE INTO civicrm_twingle_profile(name,config,last_access,access_counter) VALUES (%1, %2, null, 0)",
-        [
-          1 => [$this->name, 'String'],
-          2 => [json_encode($this->data), 'String']
-        ]);
+    try {
+      if ($this->id !== NULL) {
+        // existing profile -> just update the config
+        CRM_Core_DAO::executeQuery(
+          "UPDATE civicrm_twingle_profile SET config = %2 WHERE id = %1",
+          [
+            1 => [$this->id, 'String'],
+            2 => [json_encode($this->data), 'String'],
+          ]);
+      }
+      else {
+        // new profile -> add new entry to the DB
+        CRM_Core_DAO::executeQuery(
+          "INSERT IGNORE INTO civicrm_twingle_profile(name,config,last_access,access_counter) VALUES (%1, %2, null, 0)",
+          [
+            1 => [$this->name, 'String'],
+            2 => [json_encode($this->data), 'String'],
+          ]);
+      }
+    }
+    catch (Exception $e) {
+      throw new ProfileException(
+        E::ts("Could not save/update profile: %1", [1 => $e->getMessage()]),
+        ProfileException::ERROR_CODE_COULD_NOT_SAVE_PROFILE
+      );
     }
   }
 
   /**
    * Deletes the profile from the database
+   *
+   * @throws \CRM_Twingle_Exceptions_ProfileException
    */
   public function deleteProfile() {
-    CRM_Core_DAO::executeQuery("DELETE FROM civicrm_twingle_profile WHERE name = %1", [1 => [$this->name, 'String']]);
+    // Do only reset default profile
+    if ($this->getName() == 'default') {
+      try {
+        $default_profile = CRM_Twingle_Profile::createDefaultProfile();
+        $default_profile->setId($this->getId());
+        $default_profile->saveProfile();
+
+        // Reset counter
+        CRM_Core_DAO::executeQuery("UPDATE civicrm_twingle_profile SET access_counter = 0, last_access = NULL WHERE id = %1", [
+          1 => [
+            $this->id,
+            'Integer'
+          ]
+        ]);
+      } catch (Exception $e) {
+        throw new ProfileException(
+          E::ts("Could not reset default profile: %1", [1 => $e->getMessage()]),
+          ProfileException::ERROR_CODE_COULD_NOT_RESET_PROFILE
+        );
+      }
+    }
+    else {
+      try {
+        CRM_Core_DAO::executeQuery("DELETE FROM civicrm_twingle_profile WHERE id = %1", [
+          1 => [
+            $this->id,
+            'Integer'
+          ]
+        ]);
+      } catch (Exception $e) {
+        throw new ProfileException(
+          E::ts("Could not delete profile: %1", [1 => $e->getMessage()]),
+          ProfileException::ERROR_CODE_COULD_NOT_RESET_PROFILE
```

Comment:
Should this get a specific error code (`..._DELETE_PROFILE` instead of `..._RESET_PROFILE`)?

Reply:
Oh, the `ERROR_CODE_COULD_NOT_DELETE_PROFILE` already exists. Nice catch!